File: src/Python/core/steps.py
from __future__ import annotations
from typing import Callable, TypeVar, Any

T = TypeVar('T')
I = TypeVar('I')
O = TypeVar('O')

def ignore_errors(step: Callable[[T], T]) -> Callable[[T], T]:
    def _wrapped(inp: T) -> T:
        try:
            return step(inp)
        except Exception:
            return inp
    return _wrapped

def with_fallback(step: Callable[[I], O], fallback: Callable[[Exception], O]) -> Callable[[I], O]:
    def _wrapped(inp: I) -> O:
        try:
            return step(inp)
        except Exception as e:
            return fallback(e)
    return _wrapped

File: src/typescript/core/steps.ts
export type Unary<T> = (value: T) => T | Promise<T>;

export function ignoreErrors<T>(step: Unary<T>): Unary<T> {
  return async function wrapper(inputValue: T): Promise<T> {
    try {
      const result = await step(inputValue);
      return result;
    } catch {
      return inputValue;
    }
  };
}

export function withFallback<I, O>(step: (value: I) => O | Promise<O>,
                                   fallback: (error: unknown) => O | Promise<O>): (value: I) => Promise<O> {
  return async function wrapper(inputValue: I): Promise<O> {
    try {
      const result = await step(inputValue);
      return result;
    } catch (caughtError) {
      const alternative = await fallback(caughtError);
      return alternative;
    }
  };
}
